wire-runtime/src/main/java/com/squareup/wire/MessageAdapter.java

```diff
@@ -534,6 +552,26 @@ private void writeValueNoTag(WireOutput output, Object value, Datatype datatype)
 
   // Reading
 
+  /** Read an encoded message from {@code source}. */
+  public M read(BufferedSource source) throws IOException {
+    return read(new WireInput(source));
+  }
+
+  /** Read an encoded message from {@code bytes}. */
+  public M readBytes(byte[] bytes) throws IOException {
+    Buffer buffer = new Buffer();
+    M message = read(buffer.write(bytes));
+    if (!buffer.exhausted()) {
+      throw new IllegalArgumentException("");
+    }
+    return message;
+  }
+
+  /** Read an encoded message from {@code stream}. */
+  public M readStream(InputStream stream) throws IOException {
+    return read(Okio.buffer(Okio.source(stream)));
```

Comment:
Bytes could also be left in this buffer :cry: . Not sure what to do about that. `java.io` for the lose.